Add webauthn

mfrey43 opened the pull request:
Webauthn using https://simplewebauthn.dev/

Go to account -> Add current device -> sign out -> sign it -> enjoy!

dtslvr commented on apps/api/src/app/auth-device/auth-device.controller.ts:
```diff
@@ -0,0 +1,107 @@
+      );
+    }
+
+    const deletedAuthDevice = await this.authDeviceService.deleteAuthDevice(
```
What happens if deletedAuthDevice is `undefined`?

dtslvr commented on apps/api/src/app/auth-device/auth-device.controller.ts:
```diff
@@ -0,0 +1,107 @@
+      }
+    );
+    return {
+      id: deletedAuthDevice.id,
```
Do you need to return all this information?

dtslvr commented on apps/api/src/app/auth-device/auth-device.controller.ts:
```diff
@@ -0,0 +1,107 @@
+  @Get()
+  @UseGuards(AuthGuard('jwt'))
+  public async getAllAuthDevices(): Promise<AuthDeviceDto[]> {
+
```
Remove line break

dtslvr commented on apps/api/src/app/auth-device/auth-device.controller.ts:
```diff
@@ -0,0 +1,107 @@
+
+    return authDevices.map(authDevice => ({
+      id: authDevice.id,
+      createdAt: authDevice.createdAt.toISOString(),
```
Move up

dtslvr commented on apps/api/src/app/auth-device/auth-device.module.ts:
```diff
@@ -0,0 +1,22 @@
+  providers: [
+    AuthDeviceService,
+    PrismaService,
+    ConfigurationService,
```
Move up

dtslvr commented on apps/api/src/app/auth-device/auth-device.service.ts:
```diff
@@ -0,0 +1,70 @@
+  public async createAuthDevice(
+    data: Prisma.AuthDeviceCreateInput
+  ): Promise<AuthDevice> {
+
```
Remove line break

dtslvr commented on apps/api/src/app/auth-device/auth-device.service.ts:
```diff
@@ -0,0 +1,70 @@
+  public async updateAuthDevice(
+    params: {
+      where: Prisma.AuthDeviceWhereUniqueInput;
+      data: Prisma.AuthDeviceUpdateInput;
```
Move up

dtslvr commented on apps/api/src/app/auth/auth.controller.ts:
```diff
@@ -12,11 +13,15 @@ import { AuthGuard } from '@nestjs/passport';
 export class AuthController {
   public constructor(
     private readonly authService: AuthService,
+    private readonly webAuthService: WebAuthService,
```
Move down

dtslvr commented on apps/api/src/app/auth/auth.module.ts:
```diff
@@ -23,7 +25,9 @@ import { JwtStrategy } from './jwt.strategy';
-    UserService
+    UserService,
+    WebAuthService,
+    AuthDeviceService,
```
Move to the top

dtslvr commented on apps/api/src/services/configuration.service.ts:
```diff
@@ -24,7 +24,9 @@ export class ConfigurationService {
       REDIS_PORT: port({ default: 6379 }),
-      ROOT_URL: str({ default: 'http://localhost:4200' })
+      ROOT_URL: str({ default: 'http://localhost:4200' }),
+      WEB_AUTH_RP_ID: host({ default: 'localhost' }),
```
Is it possible to reuse `ROOT_URL` and remove this one?

dtslvr commented on apps/api/src/services/configuration.service.ts:
```diff
@@ -24,7 +24,9 @@ export class ConfigurationService {
-      ROOT_URL: str({ default: 'http://localhost:4200' })
+      ROOT_URL: str({ default: 'http://localhost:4200' }),
+      WEB_AUTH_RP_ID: host({ default: 'localhost' }),
+      WEB_AUTH_RP_NAME: str({ default: 'Ghostfolio' }),
```
I would not put this in an environment variable. Just hard code it

dtslvr commented on apps/client/src/app/components/auth-device-settings/auth-device-settings.component.ts:
```diff
@@ -0,0 +1,58 @@
+  selector: 'gf-auth-device-settings',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  templateUrl: './auth-device-settings.component.html',
+  styleUrls: ['./auth-device-settings.component.css']
```
css? 😄

dtslvr commented on apps/client/src/app/components/auth-device-settings/auth-device-settings.component.ts:
```diff
@@ -0,0 +1,58 @@
+  public isLoading = true;
+  public pageSize = 7;
+
+  constructor() { }
```
Set public explicitly

dtslvr commented on apps/client/src/app/components/auth-device-settings/auth-device-settings.component.ts:
```diff
@@ -0,0 +1,58 @@
+
+  constructor() { }
+
+  ngOnInit(): void {
```
Set public explicitly

dtslvr commented on apps/client/src/app/components/auth-device-settings/auth-device-settings.module.ts:
```diff
@@ -0,0 +1,30 @@
+  imports: [
+    CommonModule,
+    MatButtonModule,
+    MatInputModule,
```
I think you don't need `MatInputModule` here

dtslvr commented on apps/client/src/app/pages/auth-devices/auth-device-dialog/auth-device-dialog.component.ts:
```diff
@@ -0,0 +1,25 @@
+import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';
+import { AuthDeviceDto } from '@ghostfolio/api/app/auth-device/auth-device.dto';
+
+export interface AuthDeviceDialogParams {
```
Can you move this to `interfaces/interfaces.ts`?

dtslvr commented on apps/client/src/app/pages/auth-devices/auth-device-dialog/auth-device-dialog.component.ts:
```diff
@@ -0,0 +1,25 @@
+  ) {
+  }
+
+  ngOnInit(): void {
```
`public ngOnInit(): void {}`

dtslvr commented on apps/client/src/app/pages/auth-devices/auth-devices-page-routing.module.ts:
```diff
@@ -0,0 +1,14 @@
+import { NgModule } from '@angular/core';
```
I would move `<gf-auth-device-settings>` to `account-page.html` and remove this route

dtslvr commented on apps/client/src/app/pages/auth-devices/auth-devices-page.component.ts:
```diff
@@ -0,0 +1,115 @@
+  private unsubscribeSubject = new Subject<void>();
+
+
+  constructor(
```
public

dtslvr commented on apps/client/src/app/pages/auth-devices/auth-devices-page.component.ts:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  public startWebAuthn() {
+    this.http.get<any>(`/api/auth/webauthn/generate-attestation-options`, {})
```
Move api calls to a service

dtslvr commented on prisma/schema.prisma:
```diff
@@ -47,6 +47,20 @@ model Analytics {
+  createdAt           DateTime @default(now())
+  credentialId        Bytes
+  credentialPublicKey Bytes
+  counter             Int
```
Is the counter needed?

mfrey43 replied:
https://www.w3.org/TR/webauthn-2/#sctn-sign-counter

dtslvr commented on prisma/schema.prisma:
```diff
@@ -47,6 +47,20 @@ model Analytics {
+  credentialPublicKey Bytes
+  counter             Int
+  name                String
+  id                  String   @default(uuid())
```
Move up

dtslvr commented on prisma/schema.prisma:
```diff
@@ -99,20 +113,22 @@ model Settings {
+  alias         String?
+  Analytics     Analytics?
+  createdAt     DateTime   @default(now())
+  AuthDevice    AuthDevice[]
```
Move up

dtslvr commented on prisma/schema.prisma:
```diff
@@ -99,20 +113,22 @@ model Settings {
+  Settings      Settings?
+  thirdPartyId  String?
+  updatedAt     DateTime   @updatedAt
+  authChallenge String?
```
Move up

dtslvr reviewed (changes requested):
Hey @mfrey43

That looks promising! 😮  I have tested it locally Already some features I would not have expected in a first version 😎

a) Do you need the user id to login with webauthn? It does not have to work if you have not added the device yet. You could store the user id in the localstorage when adding the device.
b) Trigger webauthn in authGuard: is this the same as a), isn't it?
c) If the JWT comes from webauthn, it could be stored in session storage. Then the lifetime is limited to the session.

What do you think?

mfrey43 commented:
I have added a local copy of the types from simplewebauthn to resolve the compilation issues.

As an example I have changed for the token to be stored in session storage. How about we add a checkbox "Keep me signed in" when using the normal sign which would store it in localStorage again?

dtslvr commented on apps/client/src/app/services/token-storage.service.ts:
```diff
@@ -11,23 +11,23 @@ export class TokenStorageService {
 
   public signOut(): void {
-    const utmSource = window.localStorage.getItem('utm_source');
+    const utmSource = window.sessionStorage.getItem('utm_source');
```
This has to be kept in `localStorage`!

dtslvr replied:
If you delete the `localStorage`, you have to save it temporarily and add it again

mfrey43 replied:
Hmmm and if I no longer clear localStorage? It's weird because I need to keep track of the webauthn ids. Or should a sign out also clear those?

dtslvr replied:
If you don't clear it, you don't have to keep `utm_source` neither in localStorage nor sessionStorage 😃

dtslvr commented on apps/client/src/app/services/token-storage.service.ts:
```diff
@@ -11,23 +11,23 @@ export class TokenStorageService {
 
     if (utmSource) {
-      window.localStorage.setItem('utm_source', utmSource);
+      window.sessionStorage.setItem('utm_source', utmSource);
```
This has to be kept in `localStorage`!

dtslvr commented:
> I have added a local copy of the types from simplewebauthn to resolve the compilation issues.
> 
> As an example I have changed for the token to be stored in session storage. How about we add a checkbox "Keep me signed in" when using the normal sign which would store it in localStorage again?

Yes, that would be fine.

dtslvr commented:
You can run `yarn format` to fix the failing tests @mfrey43 😃

mfrey43 commented:
> You can run `yarn format` to fix the failing tests @mfrey43 😃

Petition to format with trailing commas 💩

dtslvr commented on apps/api/src/services/configuration.service.ts:
```diff
@@ -26,7 +26,8 @@ export class ConfigurationService {
       REDIS_PORT: port({ default: 6379 }),
-      ROOT_URL: str({ default: 'http://localhost:4200' })
+      ROOT_URL: str({ default: 'http://localhost:4200' }),
+      WEB_AUTH_RP_ID: host({ default: 'localhost' })
```
What should the value be in the cloud?

mfrey43 replied:
The domain name: ghostfolio.io
We may also be able to omit it as specified here: https://www.w3.org/TR/webauthn-2/#relying-party-identifier

dtslvr commented on apps/client/src/app/pages/account/account-page.html:
```diff
@@ -76,4 +76,37 @@ <h3 class="mb-3 text-center" i18n>Granted Access</h3>
+                i18n
+                mat-flat-button
+                (click)="startWebAuthn()"
+                [disabled]="webAuthnService.isEnabled()"
```
Nice! 🥳

dtslvr commented on apps/client/src/app/components/login-with-access-token-dialog/login-with-access-token-dialog.html:
```diff
@@ -19,6 +19,7 @@ <h1 mat-dialog-title i18n>Sign in</h1>
         [(ngModel)]="data.accessToken"
       ></textarea>
     </mat-form-field>
+    <mat-checkbox [(ngModel)]="data.staySignedIn">Stay signed in</mat-checkbox>
```
Add `i18n` directive
